### ISSUE TITLE
[FEATURE] Several FS games installed

### ISSUE THREAD

**Issue by KITT3000** [FSGModding/FSG_Mod_Assistant#69]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Is your feature request related to a problem? Please describe.

Installed FS 17, 19, 22, etc. = not very intuitive switching between games.

### Describe the solution you'd like

Adding in the options unlimited (or limited by the number of released games) the option of assigning gameSettings.xml for each FS game. Adding icons/bookmarks to the main screen with the ability to quickly switch between games (automatic change of gameSettings.xml and mod list and path to launch the game). Currently, for less experienced users, switching between FS22 and 19 is hardly intuitive (it took me a long time to explain how to do it).

### Anything else?

I haven't tested any of them, but I found this on the net, maybe it will be helpful for visualization:
https://www.kingmods.net/en/fs22/mods/fs-mod-manager
https://fs19.net/farming-simulator-2019-mods/scripts/mod-manager-fs19-17-v3-7-4-0/
https://www.modland.net/farming-simulator-2017-mods/tutorials-cheats/fs-mod-manager-v-5.0.1.100.html

**Comment by jtsage** [comment]
Hmm.  Let me kick this around for a couple days.  I had not even considered adding support for older FS games into this when I started - I was never particularly happy with how the folder checker handled that aspect.  

But here's what my gut feeling is, and I welcome discussion on any or all points.

- This is Opt-In behavior, off by default.  New preference item `Multiple FS Versions`
- When off, nothing changes.

- When this mode is on, several new preferences appear - namely, FS19 & FS17 & FS15? game and gameSettings.xml path (I missed out on 15 & 13, I know there was a modhub for 15, but did 13 have any?)
- When this mode is on, each collection will come with an option of what game it is for
- When this mode is one, this line in the UI will have a drop down at the beginning for the game you are working with

![2023-02-04 10_44_20-FSG FS22 Mod Assistant](https://user-images.githubusercontent.com/605986/216776536-4dd878d3-ad5e-46ee-a173-19b9150ce686.png)

- Changing that line will change the pre-selected game version for adding folders, and change the list of collections displayed, and change the text/icon of the launch button, the traget of "set active", etc.

Ideas?

**Comment by jtsage** [comment]
Information I believe to be correct:

- FS2015 ModDesc descVersion: 20 - 21?
- FS2017 ModDesc descVersion: 31 - 39
- FS2019 ModDesc descVersion: 40 - 59
- FS2022 ModDesc descVersion: 60+

The 2015 one was retrieved using the scientific method of downloading the first and last mod from modhub, that range may be too small, or they skipped some numbers :)

**Comment by KITT3000** [comment]
1. Turning it off by default is a good idea, hardly anyone uses several FS games.
2. Yes, no change with disabled state.
3. ModHub was already in FS13, before there was another official site to (no longer active) where there were only links to other mod sites (unofficial).
4. Yes, there should be an option to set everything required for each version of the game, i.e. gameSettings and the path to launch the game and the collection of mods.
5. You can add vertical tabs to the left side of a given FS game (as in one of the examples given above) then switching this tab will reload all settings for that version of the game.
6. modDesc version:
11: I will update when I find the dataS.gar file ;)
13: I will update when I find the dataS.gar file ;)
15: I will update when I find the dataS.gar file ;)
17: I will update when I find the dataS.gar file ;)
19: 40 - 53
22: 60 - 72 (for now)

**Comment by jtsage** [comment]
> 4\. Yes, there should be an option to set everything required for each version of the game, i.e. gameSettings and the path to launch the game and the collection of mods.

I'm actually leaning towards leaving the "meta" collection version agnostic, but each collection folder has a version attached to it.  If I try and load/replace the *set* of collections, I'm gonna run into invalidating the modCache, or worse, trying to keep more than one.  And that is a pretty big part of how it keeps speedy.  Right now, when the app gets a "display" event, it just dumps everything it knows to the render process, there is no reason I can't filter on which version of the game you want to see.

The huge upside to this is the modhub for everything prior to 22 is now static, so while it may take some time to scrape modhub id's and version, I can pack it with the app, as they won't change.

**Comment by jtsage** [comment]
Oh, another thing.  I own 19 on pc, and can pull the icon for the pretty little launch button, but I don't have 11, 13, 15 or 17 - any chance you can pull those while you are hunting?

**Comment by KITT3000** [comment]
From what I can see, MH is no longer available until FS13, for 15, 17 and 19 can still be obtained by manually changing the `&title=fs2015` parameter in the link.

As for the icon, what resolution do you need it in because I don't know if I can get the one from the EXE file or I have to look for another one (somewhere in the game files)

**Comment by KITT3000** [comment]
It seems to me that you don't need all program functions for older parts, just checking if the mod is used on save and whether it is not duplicated/in another version, and simply activating the collection for a given part (mainly it's about easier changing the location of the folder from mods).

**Comment by jtsage** [comment]
very low. I believe that image is displayed at around 25px sq, I thing I grabbed the 32x32 one so it would appear a bit nicer on scaled displays.  If memory serves, I did grab the one direct from the game EXE.

**Comment by KITT3000** [comment]
FS icon:
11: ![FS11_laucher](https://user-images.githubusercontent.com/2678653/216781329-46e3901b-e30b-4a40-841e-3ba14fd0109e.png)
13: soon
15: soon
17: soon

**Comment by jtsage** [comment]
![2023-02-04 13_24_33-FSG FS22 Mod Assistant](https://user-images.githubusercontent.com/605986/216783406-5bd85226-bcf2-4e2b-b71b-4198a1a30b91.png)

**Comment by jtsage** [comment]
Of course, the app is locked to FS22 mode for the time being so throws that style badge for mods that are not for 22

**Comment by KITT3000** [comment]
Ok, updated modDesc versions and icons ;)

**Comment by jtsage** [comment]
> Ok, updated modDesc versions and icons ;)

Thanks.  So, I managed to find a copy of 17 and 15.  11 and 13 not so much - do you still have em, do they still work?  If so, is the path to gameSettings.xml the same pattern, and is the modsDirectoryOverride done the same?  Including some other versions here for posterity.

## FS22

```xml
<gameSettings revision="15">
    <modsDirectoryOverride active="true" directory="C:\Users\jtsag\Documents\My Games\FarmingSimulator2022\mods\fsg_realism"/>
    <!-- ... -->
</gameSettings>
```

## FS19

```xml
<gameSettings revision="15">
    <modsDirectoryOverride active="true" directory="C:/Users/jtsag/Documents/My Games/FarmingSimulator2019/mods-small"/>
    <!-- ... -->
```

## FS17

```xml
<gameSettings revision="11">
    <modsDirectoryOverride active="false" directory="C:/Temp"/>
    <!-- ... -->
</gameSettings>

## FS15

```xml
<gameSettings revision="9" timescale="16" language="0" mpLanguage="0" mpColor="1" useMiles="false" moneyUnit="1">

    <modsDirectoryOverride active="false" directory="C:/Temp" />
    <!--...-->
```

**Comment by jtsage** [comment]
So, I downloaded a fair few test mods too - and what an adventure that was for the old versions.  At this time, I think I am going to omit 9 & 11 - the mod structure is probably too different to ever come back sort of clean from the scanner, and I just don't think it's worth the effort.  But supporting all the way back to 13 isn't bad.  I've done the descVersion stuff, so it'll show the FS11 flag anyway.  Maybe the 09 flag too, as I gotta add more translation strings anyway (probably - I try not to update them over and over)

**Comment by KITT3000** [comment]
I don't have the whole game, but I checked and there was an option to overwrite the path in FS11 unchanged until today in FS09 there was no such option yet

**Comment by jtsage** [comment]
A story in 3 pictures.  Still need to work out the interface side of things, but I have a pretty good idea.  Also, there are some gameSettings.xml functions that still only work with one version of that data, but slowly getting there.

![2023-02-05 23_33_12-FSG Mod Assistant](https://user-images.githubusercontent.com/605986/216884835-8b4f797c-2aa9-4605-8b6e-89fc6b9f5825.png)
![2023-02-05 23_33_19-FSG Mod Assistant](https://user-images.githubusercontent.com/605986/216884837-278c2f0f-468d-4eb5-9df6-54320ecd7f1d.png)
![2023-02-05 23_33_24-FSG Mod Assistant](https://user-images.githubusercontent.com/605986/216884839-e15e9faf-deec-41e0-bc88-cca36c75884f.png)

I asked vitalii2011 as well, who does all the Russian translations - but is there any reason to translate the string "Farming Simulator 2022", etc? It seems that the German title is the sole exception to the rule, every other iteration of the website I try has the English name. I suppose I could just add it once, and do the numbers after the string in every case.

**Comment by KITT3000** [comment]
I don't think the name of the game is worth translating. As with the translations "this mod is for FS XX" it can be simplified to the main text "this mod is for FS" and the number added automatically without separate translations.

**Comment by jtsage** [comment]
yeah, the only reason I did those is because it was easier to do the tooltip overlay based on the full string.  Otherwise it's a specific use case for those strings that I didn't want to handle :)

**Comment by jtsage** [comment]
Next part of the story, in video format


https://user-images.githubusercontent.com/605986/217155646-c87535ce-5211-4d34-ba6e-2471cdb6d504.mp4

**Comment by KITT3000** [comment]
Nice 😮

**Comment by jtsage** [comment]
A beta of this is available as a pre-release - if all goes well it'll be in 2.1.0 sometime this weekend.

Big thing that doesn't work yet - I've not compiled the 19. 17, or 15 modhub data yet, so everything will appear as "nonModHub" for a few more days.

DOWNLOAD: https://github.com/FSGModding/FSG_Mod_Assistant/releases/tag/v2.0.9
DOCS : https://fsgmodding.github.io/FSG_Mod_Assistant/multi-version.html

**Comment by KITT3000** [comment]
I tested this version a bit and the only thing missing is loading the appropriate game LOG, currently when I have active eg. FS19, the program still loads the log from FS22

**Comment by jtsage** [comment]
Yeah, I'm still turing over in my head how I'm going to go about that.  What it *actually* does is load the last log that you specifically loaded, but if there isn't one, it guesses based on the FS22 settings.

I'm hesitant to tie it to the game settings in preferences, because that's gonna be a pain to change.  Maybe a set of small-ish game icons to try and load what it thinks the default one for that game version is?  I don't wanna lose the ability to load an arbitrary game log file either.  It's also possible, at least in 22 for sure to change what the game calls the log (and add a folder to that name, but in our tests you couldn't place it arbitrarily in a folder outside of the normal one)

**Comment by KITT3000** [comment]
I think you should set a default load path which will be overwritten when the active version of the game is changed, but can be overwritten by manually uploading the file.

**Comment by jtsage** [comment]
So, here's where I'm heading with this...

![2023-02-10 15_39_08-modAssist_main js - FSG_Mod_Assistant - Visual Studio Code](https://user-images.githubusercontent.com/605986/218193136-76f76cb7-3243-40f9-8da8-79ca0f5739ab.png)

Action #1 - You click on the folder open log button - it will stay on that log file, even persisting across app open/close until you tell it to do something else

Action #2 - You click on any of those little game icons - it will load the log.txt file that resides in the same folder as the gameSettings.xml file specified for that version. (only those versions you have set it up for will appear).  When you change the game version in the app, the selection will automatically change to the "current" game log file.

Default - Basically, action #2 above, it will attempt to guess well where the log is.

The other option is I stop persisting the log path, and it auto-changes with the game version - opening a specific file in the log window will only persist until that window closes - *BUT*, if you change the version with that window open, that would **also** instantly change the contents.

**Comment by jtsage** [comment]
Another RC available, https://github.com/FSGModding/FSG_Mod_Assistant/releases/tag/v2.0.10

**Comment by vitalii2011** [comment]
![FS15](https://user-images.githubusercontent.com/41552153/218240701-fa95d792-b0a8-469e-afb2-6acf645559fb.png)
![FS17](https://user-images.githubusercontent.com/41552153/218240705-1431fa5c-5cc1-475d-ae90-9b675a48b6de.png)

**Comment by jtsage** [comment]
Yeah, I think I agree with you on 15 & 17 - The icons I have are direct from the game - and just don't fit the modern theme anymore

**Comment by jtsage** [comment]
![multi-003](https://user-images.githubusercontent.com/605986/218241791-ae9573b7-e921-46d8-bd3f-c6f9f9b2825a.png)

Honestly, I'm not really worried about the FS13 one for the half a dozen people who might still play it *and* want multiple mod sets :)

**Comment by vitalii2011** [comment]
- These logos are very similar to those on the official covers of games on Steam https://store.steampowered.com/franchise/FarmingSimulator , that is, they should probably not raise any questions. 
- In the menu, their size is noticeably larger in height than the rest. Or is it a temporary option to view/show?

**Comment by jtsage** [comment]
> - In the menu, their size is noticeably larger in height than the rest. Or is it a temporary option to view/show?

More like I didn't notice it until you pointed it out, and I was too lazy to take them into photoshop and match the blank space padding. I'll fix that, thanks

**Comment by vitalii2011** [comment]
https://user-images.githubusercontent.com/41552153/218245569-11bed3e7-52c4-4102-9fd7-4aa930a5eaa9.mp4

**Comment by jtsage** [comment]
Huh. That's odd. Yes, it uses the descVersion - the file name is only ever used as a key to grab the mod hub link and version if it exists. It might be that it is actually working in your example, but I totally messed up the addition of tags in the build you have - only the "first" tag it finds actually gets added (of those tags added in the display step, like "new" and the "fs##" ones. I'll double check this behavior.

**Comment by jtsage** [comment]
![2023-02-11 12_17_15-FSG Mod Assistant](https://user-images.githubusercontent.com/605986/218271836-87d706a3-af88-4dc1-905c-73733fabe8a7.png)
 
Some testing, looks like your test was the results of the bug fixed in 2d981a4

**Comment by jtsage** [comment]
Closing in favor of #77 - thank you for all your bug reporting efforts!
